### ISSUE TITLE
node.js may crash on windows 10 when console in/out is redirected

### ISSUE THREAD

**Issue by ikappaki** [anseki/readline-sync#94]
To reproduce, open a command prompt window and redirect input from an
empty file to a KeyIn function, e.g.

`: node -e "require('./node_modules/readline-sync/lib/readline-sync.js').keyInYN('t')" < empty.txt`

node.js crases as per exception at the bottom of this report.

The reason of the crash appears to be due to the direct use from lib/readline-sync.js:_readlineSync() of the "fs" c++ interface, and in particular of Open(), which seems to
requiring five arguments instead of three. The corresponding node.js c++
function called is src/node_file.cc:Open().

The crash is due to the following safeguard, which expects five
arguments to dispatch the synchronous call:

**src/node_file.cc : Open()**
```
 FSReqBase* req_wrap_async = GetReqWrap(args, 3);
  if (req_wrap_async != nullptr) {  // open(path, flags, mode, req)
    AsyncCall(env, req_wrap_async, args, "open", UTF8, AfterInteger,
              uv_fs_open, *path, flags, mode);
  } else {  // open(path, flags, mode, undefined, ctx)
    CHECK_EQ(argc, 5);         // ************************
    FSReqWrapSync req_wrap_sync;
    FS_SYNC_TRACE_BEGIN(open);
    int result = SyncCall(env, args[4], &req_wrap_sync, "open",
                          uv_fs_open, *path, flags, mode);
    FS_SYNC_TRACE_END(open);
    args.GetReturnValue().Set(result);
  }
```

The crash can be avoid by passing two empty additional arguments to the open function:

e.g. 

going from

`: fdR = getFsB().open('CONIN$', constants.O_RDWR, parseInt('0666', 8));`

to

`: fdR = getFsB().open('CONIN$', constants.O_RDWR, parseInt('0666', 8), null, {});`

Although thiis solves the crashing, the input redirection will not
work because 'CONIN$' is treated as a relative file, and will be
expanded to an absolute path, resulting in an attempt to open a regular file at
e.g. c:\<path>\CONIN$.

A correct solution seems to be using the fs module for grabbing the console:

e.g.

`: fdR = fs.openSync(Buffer.from('CONIN$'), "r");`

which also apperas to eradicate the need for the use of the getFsB()  function (hack?), making it more resistant to future API changes.

I will raise a pull request to discuss this as a possible fix.

Thanks

```
[4784]: c:\ws\src\node_file.cc:1550: Assertion `(argc) == (5)' failed.
 1: 00007FF6F546D1EF napi_wrap+113103
 2: 00007FF6F540CA66 v8::base::CPU::has_sse+64982
 3: 00007FF6F540CDE1 v8::base::CPU::has_sse+65873
 4: 00007FF6F53FF961 v8::base::CPU::has_sse+11473
 5: 00007FF6F5BDD850 v8::internal::Builtins::builtin_handle+323456
 6: 00007FF6F5BDCD97 v8::internal::Builtins::builtin_handle+320711
 7: 00007FF6F5BDD0D8 v8::internal::Builtins::builtin_handle+321544
 8: 00007FF6F5BDCEDE v8::internal::Builtins::builtin_handle+321038
 9: 00007FF6F6076D4D v8::internal::SetupIsolateDelegate::SetupHeap+546893
10: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
11: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
12: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
13: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
14: 00007FF6F5FF4F2C v8::internal::SetupIsolateDelegate::SetupHeap+14892
15: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
16: 00007FF6F5FF4F2C v8::internal::SetupIsolateDelegate::SetupHeap+14892
17: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
18: 00007FF6F5FF9131 v8::internal::SetupIsolateDelegate::SetupHeap+31793
19: 00007FF6F5FF8D1C v8::internal::SetupIsolateDelegate::SetupHeap+30748
20: 00007FF6F5B378A3 v8::internal::Execution::CallWasm+1395
21: 00007FF6F5B37226 v8::internal::Execution::Call+182
22: 00007FF6F5C2A4E7 v8::Script::Run+711
23: 00007FF6F54169E1 v8::base::CPU::has_sse+105809
24: 00007FF6F54196F0 v8::base::CPU::has_sse+117344
25: 00007FF6F5BDD850 v8::internal::Builtins::builtin_handle+323456
26: 00007FF6F5BDCD97 v8::internal::Builtins::builtin_handle+320711
27: 00007FF6F5BDD0D8 v8::internal::Builtins::builtin_handle+321544
28: 00007FF6F5BDCEDE v8::internal::Builtins::builtin_handle+321038
29: 00007FF6F6076D4D v8::internal::SetupIsolateDelegate::SetupHeap+546893
30: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
31: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
32: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
33: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
34: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
35: 00007FF6F5FFBBFC v8::internal::SetupIsolateDelegate::SetupHeap+42748
36: 00007FF6F5FF9131 v8::internal::SetupIsolateDelegate::SetupHeap+31793
37: 00007FF6F5FF8D1C v8::internal::SetupIsolateDelegate::SetupHeap+30748
38: 00007FF6F5B378A3 v8::internal::Execution::CallWasm+1395
39: 00007FF6F5B37226 v8::internal::Execution::Call+182
40: 00007FF6F5C0829D v8::Function::Call+605
41: 00007FF6F543679E node::Start+1150
42: 00007FF6F54369D7 node::Start+1719
43: 00007FF6F54350BA node::LoadEnvironment+26
44: 00007FF6F53C479C v8::internal::interpreter::BytecodeArrayWriter::source_position_table_builder+31596
45: 00007FF6F5436433 node::Start+275
46: 00007FF6F52A6A9C RC4_options+340380
47: 00007FF6F612ED38 v8::internal::SetupIsolateDelegate::SetupHeap+1300536
48: 00007FFD5AF27BD4 BaseThreadInitThunk+20
49: 00007FFD5B6ECE51 RtlUserThreadStart+33
```

**Comment by anseki** [comment]
Hi @ikappaki, thank you for the comment.
It seems that the `fs.c` was changed as you said.
When I wrote first version of the library, I chose `fs.c` to access to the TTY device instead of `fs` module because the `fs` didn't work for that.
However, the library is old now, and that situation changed maybe. The changing includes that `async`/`await` were added by Node.js.
Then, I think that the library is no longer required. `async`/`await` may be good choice rather than that examine the `fs.c` and those.
Anyway, I welcome comments from users for a while.
